File: app/src/main/java/com/jgdigital/ocr/MainActivity.java
package com.jgdigital.ocr;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.view.View;
import android.widget.Button;
import android.widget.ExpandableListView;
import android.widget.Toast;

import com.theartofdev.edmodo.cropper.CropImage;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private static final int REQUEST_PERMISSIONS = 100;
    private static final int REQUEST_PICK_FOLDER = 101;
    private static final int REQUEST_PICK_ZIP = 102;
    private static final int REQUEST_PICK_TXT = 103;

    private Button btnSelectFolder, btnSelectZip, btnSelectTxt;
    private ExpandableListView expandableListView;
    private ExpandableListAdapter listAdapter;
    private List<String> listDataHeader;
    private HashMap<String, List<String>> listDataChild;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize UI elements
        btnSelectFolder = findViewById(R.id.btnSelectFolder);
        btnSelectZip = findViewById(R.id.btnSelectZip);
        btnSelectTxt = findViewById(R.id.btnSelectTxt);
        expandableListView = findViewById(R.id.expandableListView);

        // Prepare expandable list data
        prepareListData();

        listAdapter = new ExpandableListAdapter(this, listDataHeader, listDataChild);
        expandableListView.setAdapter(listAdapter);

        // Set click listeners
        btnSelectFolder.setOnClickListener(v -> checkPermissionsAndOpenFolder());
        btnSelectZip.setOnClickListener(v -> checkPermissionsAndOpenZip());
        btnSelectTxt.setOnClickListener(v -> checkPermissionsAndOpenTxt());

        // ExpandableListView item click listener
        expandableListView.setOnChildClickListener((parent, v, groupPosition, childPosition, id) -> {
            String selectedItem = listDataChild.get(listDataHeader.get(groupPosition)).get(childPosition);
            Toast.makeText(MainActivity.this, "Selected: " + selectedItem, Toast.LENGTH_SHORT).show();
            // Here you would handle the OCR type selection
            return false;
        });
    }

    private void prepareListData() {
        listDataHeader = new ArrayList<>();
        listDataChild = new HashMap<>();

        // Adding headers
        listDataHeader.add("Extraction Type");

        // Adding child data
        List<String> extractionOptions = new ArrayList<>();
        extractionOptions.add("Extract all text in images");
        extractionOptions.add("Only text extract");
        extractionOptions.add("Only numbers extract");
        extractionOptions.add("7+ digit numbers extract");

        listDataChild.put(listDataHeader.get(0), extractionOptions);
    }

    private void checkPermissionsAndOpenFolder() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    REQUEST_PERMISSIONS);
        } else {
            openFolderPicker();
        }
    }

    private void checkPermissionsAndOpenZip() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    REQUEST_PERMISSIONS);
        } else {
            openZipPicker();
        }
    }

    private void checkPermissionsAndOpenTxt() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    REQUEST_PERMISSIONS);
        } else {
            openTxtPicker();
        }
    }

    private void openFolderPicker() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        startActivityForResult(intent, REQUEST_PICK_FOLDER);
    }

    private void openZipPicker() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("application/zip");
        startActivityForResult(intent, REQUEST_PICK_ZIP);
    }

    private void openTxtPicker() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("text/plain");
        startActivityForResult(intent, REQUEST_PICK_TXT);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            if (requestCode == REQUEST_PICK_FOLDER) {
                Uri treeUri = data.getData();
                // Process the selected folder
                Toast.makeText(this, "Folder selected: " + treeUri.toString(), Toast.LENGTH_SHORT).show();
            } else if (requestCode == REQUEST_PICK_ZIP) {
                Uri zipUri = data.getData();
                // Process the selected zip file
                Toast.makeText(this, "Zip file selected: " + zipUri.toString(), Toast.LENGTH_SHORT).show();
            } else if (requestCode == REQUEST_PICK_TXT) {
                Uri txtUri = data.getData();
                // Process the selected txt file
                Toast.makeText(this, "Text file selected: " + txtUri.toString(), Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_PERMISSIONS) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Permissions granted", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this, "Permissions denied", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/jgdigital/ocr/MainActivity.kt
package com.jgdigital.ocr

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import java.io.File

class MainActivity : AppCompatActivity() {
    private fun updateImagePreviews() {
        val container = findViewById<LinearLayout>(R.id.image_preview_container)
        container.removeAllViews()
        val size = resources.getDimensionPixelSize(android.R.dimen.app_icon_size)
        for (uri in imageUris) {
            val imageView = ImageView(this)
            imageView.layoutParams = LinearLayout.LayoutParams(size, size).apply {
                setMargins(8, 8, 8, 8)
            }
            imageView.setImageURI(uri)
            imageView.scaleType = ImageView.ScaleType.CENTER_CROP
            imageView.setBackgroundResource(android.R.drawable.picture_frame)
            container.addView(imageView)
        }
    }
    private lateinit var expandableListView: ExpandableListView
    private lateinit var adapter: ExpandableListAdapter
    private lateinit var listDataHeader: List<String>
    private lateinit var listDataChild: HashMap<String, List<String>>
    private var selectedExtractionType: Int = 0
    private val PICK_IMAGES_CODE = 1001
    private val PICK_ZIP_CODE = 1002
    private val PICK_TXT_CODE = 1003
    private var imageUris: MutableList<Uri> = mutableListOf()
    private var cropRect: android.graphics.Rect? = null
    private var outputTxtUri: Uri? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        expandableListView = findViewById(R.id.expandableListView)
        prepareListData()
        adapter = ExpandableListAdapter(this, listDataHeader, listDataChild)
        expandableListView.setAdapter(adapter)

        expandableListView.setOnChildClickListener { _, _, groupPosition, childPosition, _ ->
            selectedExtractionType = childPosition
            false
        }

        findViewById<Button>(R.id.btn_upload_images).setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "image/*"
            intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
            startActivityForResult(intent, PICK_IMAGES_CODE)
        }
        findViewById<Button>(R.id.btn_upload_zip).setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "application/zip"
            startActivityForResult(intent, PICK_ZIP_CODE)
        }
        findViewById<Button>(R.id.btn_crop_image).setOnClickListener {
            if (imageUris.isNotEmpty()) {
                val intent = Intent(this, CropActivity::class.java)
                intent.putExtra("imageUri", imageUris[0].toString())
                startActivityForResult(intent, 2001)
            } else {
                Toast.makeText(this, "No images uploaded", Toast.LENGTH_SHORT).show()
            }
        }
        findViewById<Button>(R.id.btn_apply_crop_all).setOnClickListener {
            if (cropRect != null && imageUris.isNotEmpty()) {
                batchCropImages()
            } else {
                Toast.makeText(this, "Crop area or images missing", Toast.LENGTH_SHORT).show()
            }
        }
        findViewById<Button>(R.id.btn_extract_text).setOnClickListener {
            if (imageUris.isNotEmpty()) {
                extractTextFromImages()
            } else {
                Toast.makeText(this, "No images uploaded", Toast.LENGTH_SHORT).show()
            }
        }
        findViewById<Button>(R.id.btn_fix_txt_file).setOnClickListener {
            if (outputTxtUri != null) {
                fixTxtFileFor7PlusDigits()
            } else {
                Toast.makeText(this, "No TXT file selected", Toast.LENGTH_SHORT).show()
            }
        }
    private fun batchCropImages() {
        // This is a simple synchronous implementation for demonstration. In production, use background threads.
        val croppedUris = mutableListOf<Uri>()
        for (uri in imageUris) {
            val bitmap = android.provider.MediaStore.Images.Media.getBitmap(contentResolver, uri)
            val rect = cropRect
            if (rect != null) {
                val cropped = Bitmap.createBitmap(bitmap, rect.left, rect.top, rect.width(), rect.height())
                // Save cropped bitmap to cache and get Uri
                val file = File.createTempFile("cropped_", ".jpg", cacheDir)
                val out = java.io.FileOutputStream(file)
                cropped.compress(Bitmap.CompressFormat.JPEG, 100, out)
                out.flush()
                out.close()
                croppedUris.add(Uri.fromFile(file))
            }
        }
        imageUris = croppedUris
        Toast.makeText(this, "Batch cropping done", Toast.LENGTH_SHORT).show()
    }

    private fun extractTextFromImages() {
        // This is a simple synchronous implementation for demonstration. In production, use background threads.
        val results = mutableListOf<Pair<String, String>>()
        for (uri in imageUris) {
            val bitmap = android.provider.MediaStore.Images.Media.getBitmap(contentResolver, uri)
            val fileName = FileUtils.getFileName(this, uri) ?: "image"
            var extractedText = ""
            val latch = java.util.concurrent.CountDownLatch(1)
            OcrUtils.recognizeText(this, bitmap) { text ->
                extractedText = filterTextByType(text, selectedExtractionType)
                results.add(Pair(fileName, extractedText))
                latch.countDown()
            }
            latch.await()
        }
        // Sort by file name
        val sorted = results.sortedBy { it.first }
        // Output to file
        val outputPath = findViewById<EditText>(R.id.et_output_path).text.toString()
        val outputFileName = findViewById<EditText>(R.id.et_output_file_name).text.toString().ifEmpty { "output.txt" }
        val file = File(outputPath, outputFileName)
        file.printWriter().use { out ->
            for ((name, text) in sorted) {
                out.println("$name\t$text")
            }
        }
        Toast.makeText(this, "Extraction done. Output: ${file.absolutePath}", Toast.LENGTH_LONG).show()
    }

    private fun filterTextByType(text: String, type: Int): String {
        return when (type) {
            0 -> text // All text
            1 -> text.replace("[0-9]+".toRegex(), "") // Only text
            2 -> text.replace("[^0-9]+".toRegex(), " ").trim() // Only numbers
            3 -> text.split("\\s+", "\n").filter { it.matches("\\d{7,}".toRegex()) }.joinToString(" ") // 7+ digit numbers
            else -> text
        }
    }

    private fun fixTxtFileFor7PlusDigits() {
        // Read the txt file, extract numbers with 7+ digits, and output to a new file
        val inputStream = contentResolver.openInputStream(outputTxtUri!!)
        val text = inputStream?.bufferedReader()?.use { it.readText() } ?: ""
        val numbers = text.split("\\s+", "\n").filter { it.matches("\\d{7,}".toRegex()) }
        val outputPath = findViewById<EditText>(R.id.et_output_path).text.toString()
        val outputFileName = findViewById<EditText>(R.id.et_output_file_name).text.toString().ifEmpty { "fixed_output.txt" }
        val file = File(outputPath, outputFileName)
        file.printWriter().use { out ->
            for (num in numbers) {
                out.println(num)
            }
        }
        Toast.makeText(this, "TXT fix done. Output: ${file.absolutePath}", Toast.LENGTH_LONG).show()
    }
        findViewById<Button>(R.id.btn_select_txt_file).setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "text/plain"
            startActivityForResult(intent, PICK_TXT_CODE)
        }
    }

    private fun prepareListData() {
        listDataHeader = listOf(getString(R.string.extract_text))
        listDataChild = hashMapOf(
            listDataHeader[0] to listOf(
                getString(R.string.extract_all),
                getString(R.string.extract_text_only),
                getString(R.string.extract_numbers_only),
                getString(R.string.extract_7plus_digits)
            )
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            when (requestCode) {
                PICK_IMAGES_CODE -> {
                    imageUris.clear()
                    data?.let {
                        if (it.clipData != null) {
                            val count = it.clipData!!.itemCount
                            for (i in 0 until count) {
                                val imageUri = it.clipData!!.getItemAt(i).uri
                                imageUris.add(imageUri)
                            }
                        } else if (it.data != null) {
                            imageUris.add(it.data!!)
                        }
                    }
                    Toast.makeText(this, "${imageUris.size} images selected", Toast.LENGTH_SHORT).show()
                    updateImagePreviews()
                }
                PICK_ZIP_CODE -> {
                    // TODO: Unzip and add image URIs
                    Toast.makeText(this, "ZIP upload not yet implemented", Toast.LENGTH_SHORT).show()
                }
                2001 -> {
                    // Crop result from CropActivity
                    val rect = data?.getParcelableExtra<android.graphics.Rect>("cropRect")
                    cropRect = rect
                    Toast.makeText(this, "Crop area set", Toast.LENGTH_SHORT).show()
                }
                PICK_TXT_CODE -> {
                    outputTxtUri = data?.data
                    Toast.makeText(this, "TXT file selected", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}
